File: components/TimelineBar.cpp
#include "TimelineBar.h"
#include "../style/StyleProps.h"

#include <QHBoxLayout>
#include <QLabel>
#include <QSlider>
#include <QElapsedTimer>
#include <QMouseEvent>
#include <QStyle>
#include <QStyleOptionSlider>


namespace {
constexpr qint64 kScrubThrottleMs = 33; // ~30Hz

// Click-to-seek slider (keeps your "real player" feel).
class ClickSeekSlider final : public QSlider {
public:
  explicit ClickSeekSlider(Qt::Orientation o, QWidget* parent = nullptr)
    : QSlider(o, parent) {}

protected:
  void mousePressEvent(QMouseEvent* e) override {
    if (e->button() == Qt::LeftButton) {
      QStyleOptionSlider opt;
      initStyleOption(&opt);

      const QRect handleRect =
        style()->subControlRect(QStyle::CC_Slider, &opt, QStyle::SC_SliderHandle, this);

      const QPoint p = e->position().toPoint();
      const bool clickedHandle = handleRect.contains(p);

      if (!clickedHandle) {
        const int pos = (orientation() == Qt::Horizontal) ? p.x() : p.y();
        const int span = (orientation() == Qt::Horizontal) ? width() : height();
        const int v = QStyle::sliderValueFromPosition(minimum(), maximum(), pos, span);
        setValue(v);
        // Let the default behavior continue; we'll treat this as an immediate scrub-release.
      }
    }
    QSlider::mousePressEvent(e);
  }
};
} // namespace

TimelineBar::TimelineBar(QWidget* parent)
  : QWidget(parent) {
  buildUi();
  wireSignals();
  reset();
}

void TimelineBar::buildUi() {
  auto* layout = new QHBoxLayout(this);
  layout->setContentsMargins(0, 0, 0, 0);
  layout->setSpacing(10);

  slider_ = new ClickSeekSlider(Qt::Horizontal, this);
  slider_->setObjectName("TimelineSlider");
  slider_->setRange(0, 0);
  slider_->setSingleStep(50);
  slider_->setPageStep(1000);
  slider_->setTracking(false); // you already found this feels better

  label_ = new QLabel("00:00.000 / 00:00.000", this);
  Style::setRole(label_, "muted");
  label_->setMinimumWidth(160);
  label_->setAlignment(Qt::AlignRight | Qt::AlignVCenter);

  layout->addWidget(slider_, 1);
  layout->addWidget(label_);
}

void TimelineBar::wireSignals() {
  static QElapsedTimer t;
  t.invalidate();

  connect(slider_, &QSlider::sliderPressed, this, [this]() {
    isScrubbing_ = true;
    emit scrubStarted();
  });

  connect(slider_, &QSlider::sliderMoved, this, [this](int value) {
    updateLabel(value, durationMs_);

    if (!enableLiveScrubSeek_) return;

    if (!t.isValid()) t.start();
    if (t.elapsed() >= kScrubThrottleMs) {
      emit scrubSeekTo(static_cast<qint64>(value));
      t.restart();
    }
  });

  connect(slider_, &QSlider::sliderReleased, this, [this]() {
    isScrubbing_ = false;
    emit scrubFinished(static_cast<qint64>(slider_->value()));
  });

  // If user clicks somewhere (not dragging), sliderMoved might not fire.
  connect(slider_, &QSlider::valueChanged, this, [this](int v) {
    if (isScrubbing_) return;
    // Keep label in sync even for click-to-seek
    updateLabel(v, durationMs_);
  });
}

void TimelineBar::reset() {
  durationMs_ = 0;
  isScrubbing_ = false;
  slider_->setRange(0, 0);
  slider_->setValue(0);
  slider_->setEnabled(false);
  updateLabel(0, 0);
}

void TimelineBar::setEnabledForMedia(bool on) {
  slider_->setEnabled(on && durationMs_ > 0);
}

void TimelineBar::setDurationMs(qint64 durMs) {
  durationMs_ = std::max<qint64>(0, durMs);
  slider_->setRange(0, static_cast<int>(durationMs_));
  slider_->setEnabled(durationMs_ > 0);
  updateLabel(slider_->value(), durationMs_);
}

void TimelineBar::setPositionMs(qint64 posMs) {
  posMs = std::max<qint64>(0, std::min(posMs, durationMs_));
  if (!isScrubbing_) {
    slider_->setValue(static_cast<int>(posMs));
  }
  updateLabel(posMs, durationMs_);
}

void TimelineBar::updateLabel(qint64 posMs, qint64 durMs) {
  label_->setText(QString("%1 / %2").arg(formatMs(posMs), formatMs(durMs)));
}

QString TimelineBar::formatMs(qint64 ms) {
  ms = std::max<qint64>(0, ms);
  const qint64 minutes = ms / 60000;
  ms %= 60000;
  const qint64 seconds = ms / 1000;
  const qint64 millis  = ms % 1000;

  return QString("%1:%2.%3")
    .arg(minutes, 2, 10, QChar('0'))
    .arg(seconds, 2, 10, QChar('0'))
    .arg(millis,  3, 10, QChar('0'));
}

File: AVA_V02_py/components/timeline_bar.py
"""Timeline slider + timestamp label for video scrubbing."""

from qt_compat import QtCore, QtGui, QtWidgets, Signal
from styles import set_role


def _format_ms(ms: int) -> str:
    ms = max(0, ms)
    minutes = ms // 60000
    ms %= 60000
    seconds = ms // 1000
    millis = ms % 1000
    return f"{minutes:02d}:{seconds:02d}.{millis:03d}"


class _ClickSeekSlider(QtWidgets.QSlider):
    def mousePressEvent(self, event: QtGui.QMouseEvent) -> None:  # noqa: N802
        if event.button() == QtCore.Qt.MouseButton.LeftButton:
            pos = event.position().toPoint()
            span = self.width() if self.orientation() == QtCore.Qt.Orientation.Horizontal else self.height()
            raw = pos.x() if self.orientation() == QtCore.Qt.Orientation.Horizontal else pos.y()
            value = QtWidgets.QStyle.sliderValueFromPosition(self.minimum(), self.maximum(), raw, span)
            self.setValue(value)
        super().mousePressEvent(event)


class TimelineBar(QtWidgets.QWidget):
    scrub_started = Signal()
    scrub_seek_to = Signal(int)
    scrub_finished = Signal(int)

    def __init__(self, parent: QtWidgets.QWidget | None = None) -> None:
        super().__init__(parent)
        self._duration_ms = 0
        self._is_scrubbing = False
        self._enable_live_scrub_seek = True
        self._scrub_clock = QtCore.QElapsedTimer()
        self._build_ui()
        self._wire_signals()
        self.reset()

    def _build_ui(self) -> None:
        layout = QtWidgets.QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(10)

        self.slider = _ClickSeekSlider(QtCore.Qt.Orientation.Horizontal, self)
        self.slider.setObjectName("TimelineSlider")
        self.slider.setRange(0, 0)
        self.slider.setSingleStep(50)
        self.slider.setPageStep(1000)
        self.slider.setTracking(False)

        self.label = QtWidgets.QLabel("00:00.000 / 00:00.000", self)
        self.label.setMinimumWidth(160)
        self.label.setAlignment(QtCore.Qt.AlignmentFlag.AlignRight | QtCore.Qt.AlignmentFlag.AlignVCenter)
        set_role(self.label, "muted")

        layout.addWidget(self.slider, 1)
        layout.addWidget(self.label)

    def _wire_signals(self) -> None:
        self.slider.sliderPressed.connect(self._on_scrub_started)
        self.slider.sliderMoved.connect(self._on_scrub_moved)
        self.slider.sliderReleased.connect(self._on_scrub_released)
        self.slider.valueChanged.connect(self._on_value_changed)

    def reset(self) -> None:
        self._duration_ms = 0
        self._is_scrubbing = False
        self.slider.setRange(0, 0)
        self.slider.setValue(0)
        self.slider.setEnabled(False)
        self._update_label(0, 0)

    def set_enabled_for_media(self, on: bool) -> None:
        self.slider.setEnabled(on and self._duration_ms > 0)

    def set_duration_ms(self, duration_ms: int) -> None:
        self._duration_ms = max(0, duration_ms)
        self.slider.setRange(0, self._duration_ms)
        self.slider.setEnabled(self._duration_ms > 0)
        self._update_label(self.slider.value(), self._duration_ms)

    def set_position_ms(self, pos_ms: int) -> None:
        pos_ms = max(0, min(pos_ms, self._duration_ms))
        if not self._is_scrubbing:
            self.slider.setValue(pos_ms)
        self._update_label(pos_ms, self._duration_ms)

    def _on_scrub_started(self) -> None:
        self._is_scrubbing = True
        self.scrub_started.emit()

    def _on_scrub_moved(self, value: int) -> None:
        self._update_label(value, self._duration_ms)
        if not self._enable_live_scrub_seek:
            return
        if not self._scrub_clock.isValid():
            self._scrub_clock.start()
        if self._scrub_clock.elapsed() >= 33:
            self.scrub_seek_to.emit(value)
            self._scrub_clock.restart()

    def _on_scrub_released(self) -> None:
        self._is_scrubbing = False
        self.scrub_finished.emit(self.slider.value())

    def _on_value_changed(self, value: int) -> None:
        if self._is_scrubbing:
            return
        self._update_label(value, self._duration_ms)

    def _update_label(self, pos_ms: int, duration_ms: int) -> None:
        self.label.setText(f"{_format_ms(pos_ms)} / {_format_ms(duration_ms)}")
